packages/atclient/src/atclient.c

```diff
@@ -239,3 +245,144 @@ void atclient_free(atclient *ctx) {
   atclient_connection_free(&(ctx->root_connection));
   atclient_connection_free(&(ctx->secondary_connection));
 }
+
+int atclient_get_encryption_key_shared_by_me(atclient *ctx, atclient_atsign *recipient, char *enc_key_shared_by_me)
+{
+    //  atclient_atkeys atkeys, const char *myatsign
+    int ret = 1;
+
+    // llookup:shared_key.recipient_atsign@myatsign
+    char *command_prefix = "llookup:shared_key.";
+
+    char *command = (char *)malloc(strlen(command_prefix) + (strlen(ctx->atsign.atsign)*2 - 1) + 3);
+    strcpy(command, command_prefix);
+    strcat(command, recipient->without_prefix_str);
+    strcat(command, ctx->atsign.atsign);
+    strcat(command, "\r\n");
+
+    const unsigned long recvlen = 1024;
+    unsigned char *recv = (unsigned char *)malloc(sizeof(unsigned char) * recvlen);
+    memset(recv, 0, sizeof(unsigned char) * recvlen);
+    unsigned long olen = 0;
+    ret = atclient_connection_send(&(ctx->secondary_connection), command, strlen((char *)command), recv, recvlen, &olen);
+    if (ret != 0)
+    {
+        return ret;
+    }
+
+    char *response = recv;
+
+    // Truncate response: "@" + myatsign + "@"
+    int response_prefix_len = strlen(ctx->atsign.without_prefix_str) + 3;
+    char *response_prefix = (char *)malloc(response_prefix_len);
+    strcpy(response_prefix, "@");
+    strcat(response_prefix, ctx->atsign.without_prefix_str);
+    strcat(response_prefix, "@");
+
+    if (atclient_stringutils_starts_with(response, recvlen, response_prefix, response_prefix_len))
+    {
+        response = response + response_prefix_len;
+    }
+
+    if (atclient_stringutils_ends_with(response, recvlen, response_prefix, response_prefix_len))
+    {
+        response[strlen(response) - strlen(response_prefix) - 1] = '\0';
+    }
+
+    // does my atSign already have the recipient's shared key?
+    if (atclient_stringutils_starts_with(response, recvlen, "data:", strlen("data:")))
+    {
+
+        response = response + 5;
+
+        // 44 + 1
+        unsigned long plaintextlen = 45;
+        unsigned char *plaintext = malloc(sizeof(unsigned char) * plaintextlen);
+        memset(plaintext, 0, plaintextlen);
+        unsigned long plaintextolen = 0;
+
+        ret = atchops_rsa_decrypt(ctx->atkeys.encryptprivatekey, (const unsigned char *)response, strlen((char *)response), plaintext, plaintextlen, &plaintextolen);
+        if (ret != 0)
+        {
+            printf("atchops_rsa_decrypt (failed): %d\n", ret);
+            return ret;
+        }
+        memcpy(enc_key_shared_by_me, plaintext, plaintextlen);
+    }
+    
+    else if (atclient_stringutils_starts_with(recv, recvlen, "error:AT0015-key not found", strlen("error:AT0015-key not found")))
+    {
+        // TODO: or do I need to create, store and share a new shared key?
+    }
+    return 0;
+}
```

Comment:
Both those pieces of memory I called out are never freed, nor passed to something else